server/repository/src/db_diesel/goods_received_line_row.rs

```diff
@@ -0,0 +1,168 @@
+pub struct GoodsReceivedLineRow {
+    pub id: String,
+    pub goods_received_id: String,
+    pub purchase_order_id: String,
+    pub received_pack_size: f64,
+    pub number_of_packs_received: f64,
+    pub batch: Option<String>,
+    pub weight_per_pack: Option<f64>,
+    pub expiry_date: Option<NaiveDate>,
+    pub line_number: i64,
+    pub item_link_id: String,
+    pub item_name: String,
+    pub location_id: Option<String>,
+    pub volume_per_pack: Option<f64>,
+    pub manufacturer_link_id: String,
+    // TODO map GoodsReceivedLineStatusMapping in OMS to is_authorized in OG: see issue [8647](https://github.com/msupply-foundation/open-msupply/issues/8068?issue=msupply-foundation%7Copen-msupply%7C8647)
+    pub status: GoodsReceivedLineStatus,
+    pub comment: Option<String>,
+}
+
+#[derive(DbEnum, Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[cfg_attr(test, derive(strum::EnumIter))]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+#[DbValueStyle = "SCREAMING_SNAKE_CASE"]
+pub enum GoodsReceivedLineStatus {
+    #[default]
+    Unauthorised,
+    Authorised,
+}
+
+pub struct GoodsReceivedLineRowRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> GoodsReceivedLineRowRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        GoodsReceivedLineRowRepository { connection }
+    }
+
+    pub fn _upsert_one(&self, row: &GoodsReceivedLineRow) -> Result<(), RepositoryError> {
+        diesel::insert_into(goods_received_line::table)
+            .values(row)
+            .on_conflict(goods_received_line::id)
+            .do_update()
+            .set(row)
+            .execute(self.connection.lock().connection())?;
+        Ok(())
+    }
+
+    pub fn upsert_one(&self, row: &GoodsReceivedLineRow) -> Result<i64, RepositoryError> {
+        self._upsert_one(row)?;
+        self.insert_changelog(row.to_owned(), RowActionType::Upsert)
+    }
+
+    fn insert_changelog(
+        &self,
+        row: GoodsReceivedLineRow,
+        action: RowActionType,
+    ) -> Result<i64, RepositoryError> {
+        // TODO NIT query goods_received for store_id instead of purchase_order once PR 8631 merged (https://github.com/msupply-foundation/open-msupply/pull/8631)
+        let purchase_order = PurchaseOrderRowRepository::new(self.connection)
+            .find_one_by_id(&row.purchase_order_id)?;
+        let store_id = match purchase_order {
+            Some(purchase_order) => purchase_order.store_id,
+            None => return Err(RepositoryError::NotFound),
+        };
+
+        let row = ChangeLogInsertRow {
+            table_name: ChangelogTableName::GoodsReceivedLine,
+            record_id: row.id,
+            row_action: action,
+            store_id: Some(store_id),
+            name_link_id: None,
+        };
+        ChangelogRepository::new(self.connection).insert(&row)
+    }
+
+    pub fn find_all(&self) -> Result<Vec<GoodsReceivedLineRow>, RepositoryError> {
```

Comment:
Don't think we'll need this function can remove. Probably the same for goods_recieved, but that's not this PR's job :)